<img src="images/usm.png" width="480" height="240" align="left"/>

# MAT281 - 2° Semestre 2019
## Aplicaciones de la Matemática en la Ingeniería

EL siguiente laboratorio debe realizarse de manera individual. No es posible copiar y pegar código de otros estudiantes. 
Pueden consultar internet (stackoverflow, por ejemplo), pero deben indicar sus fuentes cuando corresponda.

**Nombre**: Joaquín Silva

**Rol**: 201604611-9

**Fecha de entrega**: Lunes 25 Noviembre 2019.

Esta tarea puede completarse en `jupyter notebook` o `jupyter lab`, y sólo requiere la librería `pandas`. Es posible utilizar el ambiente `mat281_modulo2`.

## Tema 1 - Educated Guessing - Evolución de medios de almacenamiento [20 puntos]

![images/cd_dvd_bluray.jpg](images/cd_dvd_bluray.jpg)

Imagen obtenida de `http://rodhzuniga.blogspot.com/2013/01/diferencias-entre-cddvd-y-blu-ray.html` (revisar después de completada la tarea) - algunos derechos probablemente reservados.

Estime las siguientes cantidades aplicando la regla $1 - \pi - 10$. (Para más información, lea el siguiente artículo: [Regla $1 - \pi - 10$](https://en.wikipedia.org/wiki/Mathematical_coincidence))

Realice de manera honesta el ejercicio con una estimación con los conocimientos preliminares, obteniendo una primera estimación. ***A continuación***, busque información, indicando sus fuentes, para mejorar sus hipótesis. ¿Cuánto cambia su respuesta?

En todos los casos, suponga el bit perfectamente cuadrado e indique el lado y el área. De ser posible y conocido, compare con algún objeto.

**Pregunta 1.1.- ¿Qué tamaño tiene un bit en un CD-ROM)?** [5 puntos]

**R:** Las medidas de un CD son: 2[cm] el radio interior y 6[cm] el radio exterior. Entonces el área aproximada es:

$$A\approx 3\cdot (6^2 -2^2) = 96 [{cm}^2]$$

Entonces digamos que un CD tiene una capacidad de $700 [mb] = 700.000.000 [byte]$ y aproximando $1[byte]\approx 10[bit]$ tenemos que $700[mb]=7.000.000.000 [bit]$. Por lo tanto un bit tiene un tamaño aproximado en un CD-ROM:
$$
t_1\approx \frac{96}{7.000.000.000 } [{cm}^2/bit]\approx  1.37\cdot 10^{-8} [{cm}^2/bit]
$$


*DISCUSIÓN DE DATOS EN BASE A MEJOR INFORMACIÓN.*

Usando los valores reales tenemos que 
$$ A = \pi (6^2-2^2)=32\pi [{cm}^2]$$

Y además $700[mb]=8\cdot 700.000.000 [bit] = 5.600.000.000 [bit]$. Entonces un bit tiene una tamaño igual a (en un CD-ROM)

$$
t_1 = \frac{32\pi}{5.600.000.000 } [{cm}^2/bit] = 1.79\cdot 10^{-8} [{cm}^2/bit]
$$

Nota: El valor "real" igual es una aproximación ya que $\pi$ es irracional pero se entiende que ese es un valor muy cercano al real.

**Pregunta 1.2.- ¿Qué tamaño tiene un bit en DVD?**  [5 puntos]

Observación: un DVD tiene el mismo tamaño que un CD-ROM y considere la versión con una única capa de información. 

**R:** 

En este caso el DVD tiene una capacidad aproximada de $5[GB]$ entonces aproximando tenemos
$$
5[gb]= 5\cdot 10^{9}[byte]\approx 5\cdot 10^{10}[bit]
$$

Por lo tanto un bit tiene un tamaño aproximado en un DVD:
$$
t_2\approx \frac{96}{5\cdot 10^{10} } [{cm}^2/bit]\approx  1.92\cdot 10^{-9} [{cm}^2/bit]
$$

*DISCUSIÓN DE DATOS EN BASE A MEJOR INFORMACIÓN.*

Usando los valores reales tenemos que 
$$ A = \pi (6^2-2^2)=32\pi [{cm}^2]$$

La capacidad real de un DVD es $4.7[GB]$ entonces:
$$4.7[GB]=8\cdot 4.7\cdot 10^{9} [bit] = 3.76 \cdot 10^{10} [bit]$$

Por lo tanto un bit tiene una tamaño igual a (en un DVD)

$$
t_2 = \frac{32\pi}{3.76\cdot 10^{10}} [{cm}^2/bit] = 2.67\cdot 10^{-9} [{cm}^2/bit]
$$

**Pregunta 1.3.- ¿Qué tamaño tiene un bit en Blu-ray?** [5 puntos]

Observación: un Blu-ray tiene el mismo tamaño que un DVD o CD-ROM y considere la versión con una única capa de información. 

**R:** 

En este caso un Blu-ray tiene una capacidad aproximada de $25[GB]$ entonces aproximando tenemos
$$
25[gb]= 2.5\cdot 10^{10}[byte]\approx 2.5\cdot 10^{11}[bit]
$$

Por lo tanto un bit tiene un tamaño aproximado en un DVD:
$$
t_3\approx \frac{96}{2.5\cdot 10^{11} } [{cm}^2/bit]\approx  3.84\cdot 10^{-10} [{cm}^2/bit]
$$

*DISCUSIÓN DE DATOS EN BASE A MEJOR INFORMACIÓN.*

Usando los valores reales tenemos que 
$$ A = \pi (6^2-2^2)=32\pi [{cm}^2]$$

La capacidad real de un Blu-ray es $25[GB]$ entonces:
$$25[GB]=8\cdot 2.5\cdot 10^{10} [bit] = 2 \cdot 10^{11} [bit]$$

Por lo tanto un bit tiene una tamaño igual a (en un Blu-ray)

$$
t_3 = \frac{32\pi}{2\cdot 10^{11}} [{cm}^2/bit] = 5.02\cdot 10^{-10} [{cm}^2/bit]
$$

**Pregunta 1.4.- Conclusión** [5 puntos]

Considerando los puntos anteriores: ¿Cuántos bits de blu-ray cabrían en un bit de DVD? ¿Cuántos bits de DVD cabrían en un bit de CD-ROM?

**R:** 

1) 

$$
\frac{t_2}{t_3}\approx 5.31
$$
Entonces caben 5 bits de blu-ray en un bit de DVD.

2)

$$
\frac{t_1}{t_2}\approx 6.7
$$
Entonces caben 6 bits de DVD en un bit de CD-ROM.

**Nota: Los cálculos los hice con calculadora y no usando el mismo Python.**

## Tema 2 - Street Fighting Math meets Feynman [20 puntos]

![images/feynman.jpg](images/feynman.jpg)

Existe una anécdota relatada por Richard Feynman que ilustra su particular humor y su rapidez mental, relatada en el libro [Surely you're joking Mr. Feynman](http://buscador.usm.cl/Record/115966) que, por cierto, es completamente recomendable y que debería ser leído por todo ingeniero y científico. En el capítulo *"Lucky numbers"*, Feynamn menciona:

*One day at Princeton I was sitting in the lounge and overheard some mathematicians talking about the series for $e^x$, which is $1 + x + x^2/2! + x^3/3!$ Each term you get by multiplying the preceding term by $x$ and dividing by the next number. For example, to get the next term after $x^4/4!$ you multiply that term by $x$ and divide by $5$. It’s very simple.
When I was a kid I was excited by series, and had played with this thing. I had computed $e$ using that series, and had seen how quickly the new terms became very small.
I mumbled something about how it was easy to calculate $e$ to any power using that series (you just substitute the power for x).*

*“Oh yeah?” they said. “Well, then what’s $e^{3.3}$?” said some joker - I think it was Tukey.*

*I say, “That’s easy. It’s $27.11$”*

*Tukey knows it isn’t so easy to compute all that in your head. “Hey! How’d you do that?”*

*Another guy says, “You know Feynman, he’s just faking it. It’s not really right.”*

*They go to get a table, and while they’re doing that, I put on a few more figures.: $27.1126$ I say.
They find it in the table. “It’s right! But how’d you do it!”*

*“I just summed the series.”*

*“Nobody can sum the series that fast. You must just happen to know that one. How about $e^3$?”
“Look,” I say. “It’s hard work! Only one a day!”*

*“Hah! It’s a fake!” they say, happily.*

*“All right,” I say, “It’s $20.085$.”*

*They look in the book as I put a few more figures on. They’re all excited now, because I got another one right.
Here are these great mathematicians of the day, puzzled at how I can compute $e$ to any power! One of them says, “He just can’t be substituting and summing—it’s too hard. There’s some trick. You couldn’t do just any old number like $e^{1.4}$.”*

*I say, “It’s hard work, but for you, OK. It’s $4.05$.”*

*As they’re looking it up, I put on a few more digits and say, “And that’s the last one for the day!” and walk out.*

*What happened was this: I happened to know three numbers—the logarithm of $10$ to the base $e$ (needed to convert numbers from base $10$ to base $e$), which is $2.3026$ (so I knew that $e$ to the $2.3$ is very close to $10$), and because of radioactivity (mean-life and half-life), I knew the $\log$ of $2$ to the base $e$, which is $0.69315$ (so I also knew that $e^{0.7}$ is nearly equal to $2$). I also knew $e$ (to the 1), which is $2.71828$.*

*The first number they gave me was $e^{3.3}$, which is $e^{2.3+1}$, that is, ten times $e$, or 27.18. While they were sweating about how I was doing it, I was correcting for the extra $.0026$, producing $2.3026$.*

*I knew I couldn’t do another one; that was sheer luck. But then the guy said $e^3$: that’s $e^{2.3}$ times $e^{0.7}$, or ten times two. So I knew it was 20-something, and while they were worrying how I did it, I adjusted for the $0.693$.*

*Now I was sure I couldn’t do another one, because the last one was again by sheer luck. But the guy said $e^{1.4}$, which is $e^{0.7}$ times itself. So all I had to do is fix up $4$ a little bit!*

### Función exponencial truncada

Proporcionaremos la siguiente función exponencial de $n$ términos para responder las siguientes preguntas.

In [1]:
import math
def exponential_n_terms(x, n):
    """
    Calcula la exponencial considerando una serie finita de términos 
    de una manera directa.
    inputs: 
    x: int or real number.
    n: number of terms to consider in the series.
    returns:
    sx: 1 + x**1/1 + x**2/2! + ... + x**n/n!
    """
    Sx = 1
    px = 1
    for i in range(1,n+1):
        px = px * x / i
        Sx += px
    return Sx

#print(exponential_n_terms(1.0, 100))

print(exponential_n_terms(3.3, 15))
print(exponential_n_terms(3, 11))

27.11262722548732
20.08410308441558


**Pregunta 2.1:** $e^{3.3} \approx 27.1126$ [10 puntos]

1.- ¿Cuántos términos habría que haber usado en la serie para llegar a ese resultado? ¿Cuántas operaciones serían necesarias considerando el "truco" mencionado por Feynman?

**R:** Tendría que haber calculado 15 términos de la serie. 

2.- ¿Cómo se realiza el cálculo mental, en su primera y segunda aproximación? Indique las ecuaciones de la manera más formal posible.

**R:** En la primera aproximación utiliza que $e^{2.3}\approx 10$, entonces $e^{3.3}=e^{1}e^{2.3}\approx e\cdot 10 \approx 27.1828$.

**Pregunta 2.2:** $e^{3} \approx 20.085$ [10 puntos]

1.- ¿Cuántos términos habría que haber usado en la serie para llegar a ese resultado?

**R:** Tendría que haber calculado 15 términos de la serie. 

2.- ¿Cómo se realiza el cálculo mental, en su primera y segunda aproximación? Indique las ecuaciones de la manera más formal posible.

**R:** En la primera aproximación utiliza que $e^{3}=e^{2.3} e^{0.7}$, usando que $e^{0.7}\approx 2$ se tiene que $e^{3}\approx 10\cdot 2 \approx 20$.


**Pregunta 2.3:** $e^{1.4} \approx 4.05$ [Opcional]

1. ¿Cuántos términos habría que haber usado en la serie para llegar a ese resultado?

**R:** *[Respuesta]*

2.- ¿Cómo se realiza el cálculo mental, en su primera y segunda aproximación? Indique las ecuaciones de la manera más formal posible.

**R:** *[Respuesta]*

## Tema 3 - Análisis de Datos - de otra dimensión [60 puntos]

![images/iwanttobelieve.jpg](images/iwanttobelieve.jpg)

Realizaremos diversas técnicas de análisis sobre un curioso set de datos: una lista completa de avistamientos de OVNIS, que incluye el dia y hora del avistamiento, y la geolocalización. Este dataset no se encuentra muy actualizado y su procedencia puede resultar dudosa, además que contiene datos hasta el año 2014, pero resultará un buen conjunto de datos para analizar.

El repositorio original se encuentra en la dirección https://github.com/planetsig/ufo-reports, aunque también se encuentra replicado en varios otros lugares, como por ejemplo [kaggle](https://www.kaggle.com/NUFORC/ufo-sightings/version/1). Usaremos el archivo que ya se encuentra procesado y limpiado, que hemos dejado en el directorio `/tarea/code/UFO-scrubber.csv`.

### A. Análisis inicial de la información

Para conocer el dataset con el que trabajaremos, resulta natural usar los métodos `head()` y `describe`. Abriremos el dataset con todas las columnas de tipo `string` para poder realizar distintas operaciones de texto sobre ellas. 

In [2]:
import pandas as pd
# Avoid the warnings
pd.options.mode.chained_assignment = None  # default='warn'

# Open the file with separator ";" and all columns as strings
df_with_na = pd.read_csv("./data/UFO_scrubbed.csv", sep=",", dtype=str)

# Fill the Nans
df = df_with_na.fillna("")

In [4]:
# See the first rows
df.head()

datetime                  city state country     shape  \
0  10/10/1949 20:30            san marcos    tx      us  cylinder   
1  10/10/1949 21:00          lackland afb    tx             light   
2  10/10/1955 17:00  chester (uk/england)            gb    circle   
3  10/10/1956 21:00                  edna    tx      us    circle   
4  10/10/1960 20:00               kaneohe    hi      us     light   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   

                                            comments date posted    latitude  \
0  This event took place in early fall around 194...   4/27/2004  29.8830556   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005    29.38421   
2  Green/Orange circular disc over Chester&#44 En...   1/21/2008        53.2   
3  My older brother and twin sister were leaving ...   1/17/2004  28.9783333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004  21.4180556   

     longitude   
0   -97.9411111  
1    -98.581082  
2     -2.916667  
3   -96.6458333  
4  -157.8036111

In [5]:
df.describe(include="all").T

count unique             top   freq
datetime              80332  69586  7/4/2010 22:00     36
city                  80332  19900         seattle    525
state                 80332     68              ca   9655
country               80332      6              us  65114
shape                 80332     30           light  16565
duration (seconds)    80332    537             300   8635
duration (hours/min)  80332   8349       5 minutes   4716
comments              80332  79998                     15
date posted           80332    317      12/12/2009   1510
latitude              80332  18445      47.6063889    581
longitude             80332  19489    -122.3308333    581

Podemos observar fácilmente que el conjunto de datos se encuentra particularmente cargado a avistamientos en E.E.U.U.:

In [6]:
df["country"].value_counts()

us    65114
       9670
ca     3000
gb     1905
au      538
de      105
Name: country, dtype: int64

Para producir estadísticas más realistas, en lo sucesivo sólo utilizaremos sólamente datos relativos a avistamientos en dicho país que probablemente se encuentran mucho mejor documentados que en los otros países del dataset.

Por otra parte, ¿Existen estados conocidos sin tener el país de eeuu?

**1. Selección de datos**

**Pregunta 4.1: ¿Cuáles son las 5 avistamientos de mayor duración (en EEUU)? ¿Cuantos días han durado dichos encuentros?** [5 puntos]

Indique el código necesario para seleccionar únicamente los avistamientos ocurridos en EEUU, y que poseen un estado (state) distinto de vacío. Almacene el resultado en la variable `df_us`.

In [7]:
df_us = df.loc()[df['country'] == 'us']

dur = list(df_us['duration (seconds)'])
ciudades = list(df_us['city'])
print(df_us.shape)

vacios = df_us[ df_us['state'] == '' ].index
df_us.drop(vacios, inplace=True)
print(df_us.shape)

(65114, 11)
(65114, 11)


In [8]:
# Vemos que todos en EEUU todos los estados (state) son distinto de vacío.

for k in range(5):
    maximo = 0
    index = []
    for i in range(len(dur)):
        if(dur[i]!="2`" and dur[i]!="8`"):
            if(float(dur[i]) > maximo):
                maximo = float(dur[i])
                index = [i]
            elif(float(dur[i]) == maximo):
                index.append(i)
    print(maximo, index)
    dur[index[0]]=0

66276000.0 [60480]
52623200.0 [52325, 56117]
52623200.0 [56117]
10526400.0 [25229, 57680]
10526400.0 [57680]


In [9]:
print("El avistamiento más largo ocurrió en", ciudades[60480],"con una duración de", round(66276000/86400), "días.")
print("Los siguientes dos avistamientos más largos ocurrieron en", ciudades[52325], "y", ciudades[56117], "con una duración de", round(52623200/86400), "días.")
print("Los siguientes dos avistamientos más largos ocurrieron en", ciudades[25229], "y", ciudades[57680], "con una duración de", round(10526400/86400), "días.")

El avistamiento más largo ocurrió en greenbrier con una duración de 767 días.
Los siguientes dos avistamientos más largos ocurrieron en finley y englewood con una duración de 609 días.
Los siguientes dos avistamientos más largos ocurrieron en menifee y virginia beach con una duración de 122 días.


### B. Exploración simple de los datos

**Pregunta 4.2: ¿Cuáles son las 5 mayores duraciones de avistamientos (en EEUU)? ¿Cuantos días han durado dichos encuentros?** [5 puntos]

Indique el código necesario para responder la pregunta y comente al respecto.

Observaciones:

* En este caso se entiende que las 5 duraciones significa que se deben obtener 5 duraciones *distintas*.
* La columna "duration (seconds)" proporciona una unidad común (segundos) pero en unos pocos casos tiene un carácter adicional innecesario (los casos, `2'` o `8'`).
* Para poder obtener la mayor duración es necesario convertir la columna adecuadamente a float.

Sugerencia: 
* Remueva el caracter innecesario usando `replace`:

```python
df_us["duration (seconds)"].str.replace('`', '')
```

* Convierta la columna adecuada a `float`, y utilice `unique` and `sorted`.

**R: Esto ya fue respondido en el ítem anterior**

**Pregunta 4.3: ¿Donde han ocurrido los 5 encuentros que han durado más tiempo (en EEUU)?** [5 puntos]

La pregunta anterior permitía obtener la duración de los eventos, pero no otras características. Para poder responder donde han ocurrido esos encuentros, resulta necesario agregar una nueva columna y ordenando apropiadamente el dataframe. En este caso se pregunta por los 5 encuentros que han durado más tiempo, por lo que podría (eventualmente) existir repetición en las duraciones de los eventos.

In [10]:
df_us['duration_in_days'] = df_us["duration (seconds)"].str.replace('`', '').astype(float)/(24*60*60)
df_us.sort_values("duration_in_days").tail(5)

datetime            city state country     shape  \
71172   8/3/2008 21:00  virginia beach    va      us  fireball   
30596   3/1/1994 01:00         menifee    ca      us   unknown   
64390  8/10/2012 21:00          finley    wa      us     light   
69215  8/24/2002 01:00       englewood    fl      us     light   
74660  9/15/1991 18:00      greenbrier    ar      us     light   

      duration (seconds) duration (hours/min)  \
71172           10526400             4 months   
30596           10526400             4 months   
64390           52623200             2 months   
69215           52623200             2 months   
74660           66276000             21 years   

                                                comments date posted  \
71172  this object was very high up and emmited no so...   8/12/2008   
30596           Sun City / Menifee UFO sightings in 1994   2/22/2005   
64390  There have been several flying objects in a pe...   8/19/2012   
69215  bright stars&#44 moving erratically&#44 over t...   8/28/2002   
74660  Orange or amber balls or orbs of light multipl...   3/31/2008   

         latitude    longitude   duration_in_days  
71172  36.8527778   -75.9783333        121.833333  
30596  33.7283333  -117.1455556        121.833333  
64390  46.1541667  -119.0327778        609.064815  
69215  26.9616667   -82.3527778        609.064815  
74660  35.2338889   -92.3875000        767.083333

**Ya fue respondido anteriormente**

### C. Manipulación de strings
Para realizar un análisis sobre los años del avistamiento, es necesario convertir la fecha `datetime` en un formato apropiado. En python existe un tipo propio llamado precisamente `datetime` que proporciona métodos para manipulación de fechas y horas, pero para nuestros fines pedagógicos seguiremos utilizando strings. Sin embargo, si lo deseas, puedes investigar y utilizar las funciones propias de dicha librería en esta tarea (pero representará probablemente mucho más trabajo).

El formato de fecha utilizado es "mm/dd/yyyy hh:mm". Podemos extraer la información de manera sencilla utilizando las funciones de strings como se muestra a continuación:

In [11]:
# Split each row using the str method split and then assign to new columns
df_us["date"], df_us["time"] = df_us["datetime"].str.split(" ").str

In [12]:
df_us["date"].head()

0    10/10/1949
3    10/10/1956
4    10/10/1960
5    10/10/1961
7    10/10/1965
Name: date, dtype: object

In [13]:
df_us["time"].head()

0    20:30
3    21:00
4    20:00
5    19:00
7    23:45
Name: time, dtype: object

Observación: Existe en el dataset una hora "24:00" que no queda completamente claro si es "00:00" del mismo día o "00:00" del día siguiente, por lo que simplemente lo dejaremos tal cual.

**Pregunta 4.4: Genera nuevas columnas year, month, day, hour y minute con el año, mes, día, hora y minuto del avistamiento.** [5 puntos]

Indique el código necesario para responder las preguntas y comente al respecto.

In [25]:
df_us['month'], df_us['day'], df_us['year'] = df_us["date"].str.split("/").str
df_us["year"].head()

0    1949
3    1956
4    1960
5    1961
7    1965
Name: year, dtype: object

In [26]:
df_us['hour'], df_us['minute'] = df_us["time"].str.split(":").str
df_us['hour'].head()

0    20
3    21
4    20
5    19
7    23
Name: hour, dtype: object

### D. Agrupación y pivoteo de datos.

**Pregunta 4.5: ¿En qué año se observaron más avistamientos? ¿En qué mes se observan más avistamientos? ¿En qué hora se realizan más avistamientos?** [5 puntos] 

Indique el código necesario para responder las preguntas y comente al respecto.

In [29]:
# Para responder estas preguntas basta usar el comando describe
df_us.describe(include="all").T

count unique             top   freq       mean      std  \
datetime              65114  57074  7/4/2010 22:00     31        NaN      NaN   
city                  65114  11364         seattle    524        NaN      NaN   
state                 65114     52              ca   8912        NaN      NaN   
country               65114      1              us  65114        NaN      NaN   
shape                 65114     29           light  13473        NaN      NaN   
duration (seconds)    65114    468             300   7105        NaN      NaN   
duration (hours/min)  65114   6850       5 minutes   3995        NaN      NaN   
comments              65114  64866                     13        NaN      NaN   
date posted           65114    315      12/12/2009   1246        NaN      NaN   
latitude              65114  11615      47.6063889    581        NaN      NaN   
longitude             65114  12430    -122.3308333    581        NaN      NaN   
duration_in_days      65114    NaN             NaN    NaN  0.0671298  4.68187   
date                  65114   9818        7/4/2010    183        NaN      NaN   
time                  65114   1356           22:00   3782        NaN      NaN   
day                   65114     31              15   4686        NaN      NaN   
month                 65114     12               7   7639        NaN      NaN   
year                  65114     83            2012   6320        NaN      NaN   
minute                65114     60              00  26596        NaN      NaN   
hour                  65114     25              21   9647        NaN      NaN   

                              min          25%         50%         75%  \
datetime                      NaN          NaN         NaN         NaN   
city                          NaN          NaN         NaN         NaN   
state                         NaN          NaN         NaN         NaN   
country                       NaN          NaN         NaN         NaN   
shape                         NaN          NaN         NaN         NaN   
duration (seconds)            NaN          NaN         NaN         NaN   
duration (hours/min)          NaN          NaN         NaN         NaN   
comments                      NaN          NaN         NaN         NaN   
date posted                   NaN          NaN         NaN         NaN   
latitude                      NaN          NaN         NaN         NaN   
longitude                     NaN          NaN         NaN         NaN   
duration_in_days      1.15741e-07  0.000347222  0.00208333  0.00694444   
date                          NaN          NaN         NaN         NaN   
time                          NaN          NaN         NaN         NaN   
day                           NaN          NaN         NaN         NaN   
month                         NaN          NaN         NaN         NaN   
year                          NaN          NaN         NaN         NaN   
minute                        NaN          NaN         NaN         NaN   
hour                          NaN          NaN         NaN         NaN   

                          max  
datetime                  NaN  
city                      NaN  
state                     NaN  
country                   NaN  
shape                     NaN  
duration (seconds)        NaN  
duration (hours/min)      NaN  
comments                  NaN  
date posted               NaN  
latitude                  NaN  
longitude                 NaN  
duration_in_days      767.083  
date                      NaN  
time                      NaN  
day                       NaN  
month                     NaN  
year                      NaN  
minute                    NaN  
hour                      NaN

1) ¿En qué año se observaron más avistamientos? 

**R: En el año 2012 con 6320 avistamientos.**

2) ¿En qué mes se observan más avistamientos?

**R: En el mes de julio (7) con 7639 avistamientos.**

3) ¿En qué hora se realizan más avistamientos?

**R: A las 21 horas con 9647 avistamientos.**

**Pregunta 4.6: ¿Cuántos avistamientos y qué porcentaje de avistamientos de tipo "platillo volador clásico" se observan por año en cada estado?** [5 puntos]

Definamos como forma *"platillo volador clásico"* a los avistamientos que poseen forma triangular, circular, esferica, ovalada o disco (triangle, circle, sphere, disk u oval). ¿Cómo podemos calcular que porcentaje de avistamientos **por año** corresponden a este tipo?

Los pasos a realizar los son siguientes:

1.- Definir una variable que permita distinguir a los patillos voladores clásicos de los otros. En general, se recomienda usar 1 para "platillo volador clasico" y 0 para otro caso, para poder usar las funciones suma y count.

2.- Realizar un groupby apropiado y realizar el conteo necesario.

3.- Realizar la estadística respectiva.

In [77]:
# 1. Definición de variable apropiada
df_us["is_flying_saucer"] = 0

k = 0
for i in df_us['shape']:
    if(i == "triangle" or i == "circle" or i == "sphere" or i == "disk" or i == "oval"):
        df_us["is_flying_saucer"].iloc[[k]]='1'
    k = k+1

In [130]:
# 2. Groupby
df_count = df_us.groupby('state')

estados = []
porcentajes = []

for i in df_count.groups:
    estado = df_count.get_group(i) 
    estados.append(i)
    suma = 0
    for k in estado['is_flying_saucer']:
            if(k == '1'):
                suma = suma+1
    total = len(list(estado['is_flying_saucer']))
    porcentajes.append(suma*100/total)    

In [131]:
for i in range(len(estados)):
    print("El estado", estados[i], "tiene", round(porcentajes[i]),"% de avistamientos clásicos.")

El estado ak tiene 39 % de avistamientos clásicos.
El estado al tiene 40 % de avistamientos clásicos.
El estado ar tiene 39 % de avistamientos clásicos.
El estado az tiene 35 % de avistamientos clásicos.
El estado ca tiene 37 % de avistamientos clásicos.
El estado co tiene 38 % de avistamientos clásicos.
El estado ct tiene 36 % de avistamientos clásicos.
El estado dc tiene 29 % de avistamientos clásicos.
El estado de tiene 50 % de avistamientos clásicos.
El estado fl tiene 36 % de avistamientos clásicos.
El estado ga tiene 39 % de avistamientos clásicos.
El estado hi tiene 37 % de avistamientos clásicos.
El estado ia tiene 38 % de avistamientos clásicos.
El estado id tiene 36 % de avistamientos clásicos.
El estado il tiene 39 % de avistamientos clásicos.
El estado in tiene 37 % de avistamientos clásicos.
El estado ks tiene 33 % de avistamientos clásicos.
El estado ky tiene 39 % de avistamientos clásicos.
El estado la tiene 40 % de avistamientos clásicos.
El estado ma tiene 40 % de avis

R: *< < RESPUESTA > >*

**Pregunta 4.7: Pivotee la tabla anterior para que se indique en cada fila, un año con la cantidad de avistamientos de tipo platillo volador para cada estado.** [5 puntos]

La tabla debería tener como columnas: `state`, `ak`, `al`, `az`, ..., `wi`, `wv`, `wy`. Utilice `fillna` para llenar con string vacío para facilitar la lectura de la tabla. 

Indique el código y comente.

In [ ]:
# R:
#df_pt = 

R: *< < RESPUESTA > >*

A continuación, realizaremos varios pasos para conocer cuál estado de EEUU tiene más avistamientos por persona por año.

Inocentemente, uno podría pensar que el mejor lugar para realizar avistamientos de ovnis es el lugar donde existen más registros de avistamientos. Sin embargo, resulta natural que existan más avistamientos en los lugares donde existe un mayor asentamiento de personas. Un indicador más representativo de la posibilidad de realizar un avistamiento es el número de observaciones por cada 10,000 personas. Sin embargo, como no tenemos la población de cada estado en esta tabla tendremos que ¡mezclar distintas tablas!

Lamentablemente no es posible conseguir fácilmente una tabla con la población de cada estado, a través de los años. Sin embargo, como nos interesan las estadísticas más recientes, podemos considerar que la proporción entre los distintos estados no ha cambiado mayormente en los últimos años, al menos en estas estadísticas iniciales. Por ende, las estadísticas proporcionadas por https://datamarket.com nos sirven como una buena referencia de población por estado. 

El archivo `data/us-population-by-state-and-count.csv` contiene la población por estado para el año 2012, el cual se encuentra en formato csv pero con separador `;`.

Abramos el archivo en un dataframe llamado `df_pop`:

In [ ]:
df_pop = pd.read_csv("data/us-population-by-state-and-count.csv", sep=";")
df_pop.head()

¡Los estados aparecen con su nombre completo y no con la abreviación como en el archivo de los OVNIS!
Afortunadamente, existe un tercer archivo que permite hacer la relación entre los archivos.

In [ ]:
df_abr = pd.read_csv("data/us-abbreviation.csv", sep=";")
df_abr.head()

Realizaremos un merge entre los dataframes `df_pop` y `df_abr` para obtener un dataframe `df_pop_abr` con cuatro columnas: state_name, state_population, State Name,y State Abbreviation. 

In [ ]:
# Merge de nuestros dataframes
df_pop_abr = df_pop.merge(df_abr, left_on=["state_name"], right_on=["State Name"], how="outer").fillna("")
df_pop_abr

**Pregunta 4.8:  Realice un poco de limpieza antes del segundo merge.**  [5 puntos]

La columna con la que haremos el merge se llama `"state"` y tiene el código de los estados en minúsculas. Para que todo funcione más facil, resulta conveniente subseleccionar las columnas y usar minúsculas en un nuevo dataframe que llamaremos `df_r` y que debería verse así:

```
state    state_name    state_population
al       Alabama       4777326
...
wy       Wyoming       562803
```


Indique el código necesario.

In [ ]:
# R: << CODIGO ACA >>
#df_pop_abr["state"] = 
#df_r = 
#df_r

R: *< < COMENTAR AQUI > >*

**Pregunta 4.9:  Calcule la cantidad de avistamientos en cada etado, durante los últimos 5 años, almacenandolo en un dataframe llamado `df_state_5y`** [5 puntos]
 
Indique el código necesario para responder la pregunta y comente al respecto.

In [ ]:
#df_5y = 
#df_state_5y = 
#df_state_5y["sightings_in_5_years"] = 
#df_state_5y

**Pregunta 4.10:  Realice un segundo merge, entre los dataframes `df_state_5y` y `df_r` para obtener un dataframe `df_sight` en el cual se agregue la columna de population.**  [5 puntos]

Indique el código necesario para responder la pregunta y comente al respecto.

In [ ]:
#df_sight = 
#df_sight

R: *< < COMENTAR AQUI > >*

**Pregunta 4.11:  Agregue una columna llamada `expected_sightings` indicando la cantidad de avistamientos esperados para 10,000 personas a lo largo de 5 años.**  [5 puntos]

Indique el código necesario para responder la pregunta y comente al respecto.

In [ ]:
#df_sight["expected_sightings"] 

R: *< < COMENTAR AQUI > >*

**Pregunta 4.12: Ordenando por la columna `expected_sightings`, ¿cuáles son los 3 mejores estados de EEUU para avistar OVNIS?. ¿Cuál otro factor, a su juicio, hubiera sido importante a considerar?**  [5 puntos]

Indique el código necesario para responder la pregunta y comente al respecto.

In [ ]:
#df_sight

R: *< < COMENTAR AQUI > >*

## Tema 4 - Retroalimentación sobre el curso [0 puntos]

Esta sección **no será evaluada** pero nos permitirá obtener retroalimentación de las clases, tareas y laboratorios. Puedes ser completamente sincero, pues no será parte de la evaluación de esta tarea (ni de futuras).

**1**. ¿Aproximadamente cuánto tiempo, en horas, te tomó terminar la tarea?

*R: << RESPUESTA DEL ALUMNO >>*

**2**. ¿Qué tan interesante es esta tarea (0-100)?

*R: << RESPUESTA DEL ALUMNO >>*

**3**. ¿Qué tan difícil fue esta tarea (0-100)? Siendo 0 complemente trivial, 100 completamente imposible y 50 siendo una complejidad apropiada.

*R: << RESPUESTA DEL ALUMNO >>*

        